In [29]:
import pandas as pd
import geopandas as gpd
import numpy as np
import re

# Residential
## Residential Buildings

In [12]:
df_resbuild = gpd.read_file('https://opendata.arcgis.com/datasets/8b57e88e62cf4d9791cf5568675d05da_0.geojson')

In [39]:
resbuild = gpd.GeoDataFrame(columns=df_resbuild.columns)
def get_number_from_range(num_range):
    if '-' not in num_range:
        return [num_range]
    numMin, numMax = num_range.split("-")
    if '/' in numMin:
        numMin = numMin[numMin.index('/')+1:]
    try:
        return [i for i in range(int(numMin), int(numMax)+1)]
    except:
        print(num_range)
for index, row in df_resbuild.iterrows():
    nums = row['Str_Number'].split(',')
    for num in nums:
        for str_num in get_number_from_range(num):
            nrow = row.copy()
            nrow['StreetNumber'] = str_num
            resbuild = resbuild.append(nrow)
resbuild.drop(['Str_Number'], axis=1, inplace=True)
resbuild.head(n=10)

FID BLDG_ID      BLDG_Name  StreetName               Address  Area_Ha  \
0   1    1855  Airways Plaza  Derry Rd E  2889-2907 Derry Rd E     0.33   
0   1    1855  Airways Plaza  Derry Rd E  2889-2907 Derry Rd E     0.33   
0   1    1855  Airways Plaza  Derry Rd E  2889-2907 Derry Rd E     0.33   
0   1    1855  Airways Plaza  Derry Rd E  2889-2907 Derry Rd E     0.33   
0   1    1855  Airways Plaza  Derry Rd E  2889-2907 Derry Rd E     0.33   
0   1    1855  Airways Plaza  Derry Rd E  2889-2907 Derry Rd E     0.33   
0   1    1855  Airways Plaza  Derry Rd E  2889-2907 Derry Rd E     0.33   
0   1    1855  Airways Plaza  Derry Rd E  2889-2907 Derry Rd E     0.33   
0   1    1855  Airways Plaza  Derry Rd E  2889-2907 Derry Rd E     0.33   
0   1    1855  Airways Plaza  Derry Rd E  2889-2907 Derry Rd E     0.33   

   Area_Acre  GFA_RES_m2  GFARes_ft2  GFA_RET_m2  ...  RES_Units  RET_Units  \
0   0.815448         0.0         0.0         0.0  ...         10         10   
0   0.815448         0.0         0.0         0.0  ...         10         10   
0   0.815448         0.0         0.0         0.0  ...         10         10   
0   0.815448         0.0         0.0         0.0  ...         10         10   
0   0.815448         0.0         0.0         0.0  ...         10         10   
0   0.815448         0.0         0.0         0.0  ...         10         10   
0   0.815448         0.0         0.0         0.0  ...         10         10   
0   0.815448         0.0         0.0         0.0  ...         10         10   
0   0.815448         0.0         0.0         0.0  ...         10         10   
0   0.815448         0.0         0.0         0.0  ...         10         10   

   Inst_Beds Units_Ha Units_Acre  FSI Year_Built WardNumber  \
0          0       61         25  0.0                     5   
0          0       61         25  0.0                     5   
0          0       61         25  0.0                     5   
0          0       61         25  0.0                     5   
0          0       61         25  0.0                     5   
0          0       61         25  0.0                     5   
0          0       61         25  0.0                     5   
0          0       61         25  0.0                     5   
0          0       61         25  0.0                     5   
0          0       61         25  0.0                     5   

                          geometry StreetNumber  
0  MULTIPOINT (-79.64386 43.70344)         2889  
0  MULTIPOINT (-79.64386 43.70344)         2890  
0  MULTIPOINT (-79.64386 43.70344)         2891  
0  MULTIPOINT (-79.64386 43.70344)         2892  
0  MULTIPOINT (-79.64386 43.70344)         2893  
0  MULTIPOINT (-79.64386 43.70344)         2894  
0  MULTIPOINT (-79.64386 43.70344)         2895  
0  MULTIPOINT (-79.64386 43.70344)         2896  
0  MULTIPOINT (-79.64386 43.70344)         2897  
0  MULTIPOINT (-79.64386 43.70344)         2898  

[10 rows x 27 columns]

In [40]:
resbuild.to_file('/Users/Nick/Documents/Personal/Data/Mississauga/residential_buildings.geojson', driver='GeoJSON')

In [8]:
df_resbuild.columns

Index(['FID', 'BLDG_ID', 'BLDG_Name', 'Str_Number', 'StreetName', 'Address',
       'Area_Ha', 'Area_Acre', 'GFA_RES_m2', 'GFARes_ft2', 'GFA_RET_m2',
       'GFARet_ft2', 'GFA_Tot_m2', 'GFATot_ft2', 'BLDG_Type', 'Tenure_RES',
       'Tenure_RET', 'Storeys', 'RES_Units', 'RET_Units', 'Inst_Beds',
       'Units_Ha', 'Units_Acre', 'FSI', 'Year_Built', 'WardNumber',
       'geometry'],
      dtype='object')

In [12]:
df_resbuild.head().iloc[1]

FID                                                       2
BLDG_ID                                                1856
BLDG_Name                                                  
Str_Number                                             2921
StreetName                                       Derry Rd E
Address                                     2921 Derry Rd E
Area_Ha                                                0.15
Area_Acre                                          0.370658
GFA_RES_m2                                                0
GFARes_ft2                                                0
GFA_RET_m2                                                0
GFARet_ft2                                                0
GFA_Tot_m2                                                0
GFATot_ft2                                                0
BLDG_Type                                     Mixed RES/RET
Tenure_RES                                                 
Tenure_RET                              

In [9]:
df_resbuild.BLDG_Type.unique()

array(['Mixed RES/RET', 'Apartment', 'Residential Institution', 'Plex'],
      dtype=object)

2921 is said to be a mixed residential retail zone, but looking on Google Maps we see the a law office and lounge. Thusly, it cannot be assumed that the type of building always maps exactly to their use, at least when talking about Mixed Use. 

<img src="./RefImages/2921DerryRoadEast.png" width="200"/>


## Townhouses

In [6]:
df_townhouses = gpd.read_file('https://opendata.arcgis.com/datasets/34a39af0d72143a7a74e675fe5fdcbbf_0.geojson')
df_townhouses.columns

Index(['FID', 'COMPLEX_ID', 'CHAR_AREA', 'Z_AREA', 'Str_Number', 'StreetName',
       'Address', 'OP_DES', 'Zoning', 'Cplex_Name', 'Area_Ha', 'Area_Acre',
       'Parking', 'GFA_RES_m2', 'GFARes_ft2', 'GFA_RET_m2', 'GFARet_ft2',
       'GFA_Tot_m2', 'GFATot_ft2', 'Cplex_Type', 'Tenure_RES', 'Storeys',
       'TH_Blocks', 'RES_Units', 'RET_Units', 'Unit_Ha', 'Unit_Acre', 'FSI',
       'Year_Built', 'WardNumber', 'Shape__Area', 'Shape__Length', 'geometry'],
      dtype='object')

In [20]:
df_townhouses.Cplex_Type.unique()

array(['Townhouse', 'Detached', 'Other (Mobile Homes)', 'Semi-Detached'],
      dtype=object)

In [21]:
df_townhouses.OP_DES.unique()

array(['Medium Density', 'Low Density II', 'High Density',
       'Institutional', 'Mixed Use', 'Low  Density I', 'Low Density I',
       'Low Density II, Medium Density', 'Medium Density, HighDensity',
       'Mixed Use, Medium Density'], dtype=object)

There are detached homes, as well as high and low density homes. Let's take a look at those homes in particular. 

In [23]:
df_townhouses[df_townhouses['Cplex_Type'] == 'Detached'].sample()

FID  COMPLEX_ID              CHAR_AREA Z_AREA      Str_Number  \
422  423        1395  Cooksville NHD (West)     15  2165,2175/2130   

                  StreetName                                 Address  \
422  Stavebank Rd/Dickson Rd  2165,2175/2130 Stavebank Rd/Dickson Rd   

             OP_DES Zoning  Cplex_Name  ...  RES_Units  RET_Units  Unit_Ha  \
422  Low  Density I  R1-19  The Colony  ...         40          0       10   

     Unit_Acre       FSI  Year_Built  WardNumber   Shape__Area  Shape__Length  \
422          4  0.223423        2001           7  76315.574219    1351.730864   

                                              geometry  
422  POLYGON ((-79.61740 43.56039, -79.61716 43.560...  

[1 rows x 33 columns]

2165 Stavebank Rd is detached and has a density level of Low Density 1. Looking on Google Maps we see a large low density detached home. 

## Residential Centres

In [8]:
df_res_centres = gpd.read_file("https://opendata.arcgis.com/datasets/04fcf2b85ba04fe5bd706939f0d811d8_0.geojson")

In [5]:
df_res_centres.columns

Index(['FID', 'CENTRE_ID', 'CHAR_AREA', 'OP_DES', 'ZONING', 'Z_AREA',
       'N_OF_BLDGS', 'PARKING', 'Shape__Area', 'Shape__Length', 'geometry'],
      dtype='object')

In [10]:
df_res_centres['OP_DES'].unique()

array(['High Density', 'Medium Density', 'Low Density I', 'Mixed Use',
       'Low Density II', 'High Density, Greenbelt', 'Institutional',
       'Low Density I, Greenbelt', 'Business Employment', 'Office',
       'Convenience Commercial', 'High Density, Mixed Use'], dtype=object)

In [14]:
df_res_centres[df_res_centres['OP_DES'] == 'Low Density I, Greenbelt']

FID  CENTRE_ID                  CHAR_AREA                    OP_DES  \
581  582       1355  Clarkson - Lorne Park NHD  Low Density I, Greenbelt   

    ZONING Z_AREA  N_OF_BLDGS  PARKING   Shape__Area  Shape__Length  \
581  R1-36      3           1       43  26778.964844    1004.267222   

                                              geometry  
581  POLYGON ((-79.62300 43.51242, -79.62297 43.512...

# Offices
## Office Buildings

In [15]:
df_office_bldgs = gpd.read_file('https://opendata.arcgis.com/datasets/439debbd5bb142c98f834095e730f831_0.geojson')
df_office_bldgs.columns

Index(['FID', 'BLDG_ID', 'BLDG_Name', 'Address', 'OfficeType', 'OFF_GFA_m2',
       'OFF_GFAft2', 'Storeys', 'Year_Built', 'geometry'],
      dtype='object')

In [16]:
df_office_bldgs['OfficeType'].unique()

array(['Major Office', 'Secondary Office'], dtype=object)

In [18]:
df_office_bldgs[df_office_bldgs['OfficeType'] == 'Major Office'].sample(n=5)

FID  BLDG_ID                         BLDG_Name              Address  \
374  375      461                                        2100 Derry Rd W   
583  584      709   SOBEYS ONTARIO  REGIONAL OFFICE      4980 Tahoe Blvd   
201  202      251  Mississauga Executive Centre III   3 Robert Speck Pky   
177  178      223        Markborough Place - Phs II  6733 Mississauga Rd   
152  153      198                                     1919 Minnesota Crt   

       OfficeType  OFF_GFA_m2     OFF_GFAft2  Storeys Year_Built  \
374  Major Office     10293.3  110796.159023      4.0  2008/2009   
583  Major Office     18934.0  203803.879703      6.0       2014   
201  Major Office     27090.0  291594.333007     15.0       1987   
177  Major Office     18460.0  198701.786169      8.0       1990   
152  Major Office     10913.6  117473.012651      5.0       2013   

                            geometry  
374  MULTIPOINT (-79.75342 43.60856)  
583  MULTIPOINT (-79.61074 43.64375)  
201  MULTIPOINT (-79.63727 43.59786)  
177  MULTIPOINT (-79.74056 43.60285)  
152  MULTIPOINT (-79.74254 43.60496)

Sobey's Ontario Regional Office, which is located on 4980 Tahoe Blvd, is a large building classified as a major office:
<img src="./RefImages/Sobeys.png" width=500/>

In [20]:
df_office_bldgs[df_office_bldgs['OfficeType'] == 'Secondary Office'].sample(n=5)

FID  BLDG_ID                         BLDG_Name              Address  \
516  517      635                                          99 Queen St S   
167  168      213                Meadowvale Court I  6535 Mississauga Rd   
72    73       95     Dixie Court Corporate Parks 4        6375 Dixie Rd   
287  288      360  Silicon Graphics Office Building     5800 Explorer Dr   
595  596      721                                       2600 Skymark Ave   

           OfficeType  OFF_GFA_m2     OFF_GFAft2  Storeys Year_Built  \
516  Secondary Office       186.0    2002.087336      2.0              
167  Secondary Office       488.0    5252.788280      1.0       1979   
72   Secondary Office      5659.0   60912.969010      4.0       1988   
287  Secondary Office      9244.5   99506.969785      5.0       1998   
595  Secondary Office      9930.0  106885.630371      2.0       1992   

                            geometry  
516  MULTIPOINT (-79.71702 43.58445)  
167  MULTIPOINT (-79.73739 43.59902)  
72   MULTIPOINT (-79.65628 43.66391)  
287  MULTIPOINT (-79.61652 43.65601)  
595  MULTIPOINT (-79.59803 43.65797)

5800 Explorer Drive, though a secondary office, is about the ssame size as the Sobey's office. The office type is thus, not a good tell of the size of the office. If we were to focus on OFF_GFA_m2 which I believe have to do with actual office size: 

In [24]:
df_office_bldgs.sort_values('OFF_GFA_m2').head()

FID  BLDG_ID BLDG_Name                 Address        OfficeType  \
507  508      625                     33 Queen St S  Secondary Office   
572  573      698            2537 Confederation Pky  Secondary Office   
460  461      567                1413 Hurontario St  Secondary Office   
563  564      685                     70 Queen St S  Secondary Office   
491  492      608                      46 Port St E  Secondary Office   

     OFF_GFA_m2  OFF_GFAft2  Storeys Year_Built  \
507       52.00  559.723341      1.0       1950   
572       72.87  784.366152      1.0       1952   
460       78.50  844.966967      1.0              
563       83.61  899.970549      1.0       1953   
491       87.00  936.460206      1.0              

                            geometry  
507  MULTIPOINT (-79.72002 43.58663)  
572  MULTIPOINT (-79.61700 43.57630)  
460  MULTIPOINT (-79.59244 43.56368)  
563  MULTIPOINT (-79.71864 43.58501)  
491  MULTIPOINT (-79.58259 43.55308)

In [25]:
df_office_bldgs.sort_values('OFF_GFA_m2').tail()

FID  BLDG_ID              BLDG_Name             Address    OfficeType  \
582  583      708                            60 Standish Crt  Major Office   
54    55       71  Mississauga City Hall  300 City Centre Dr  Major Office   
283  284      349           Ingram Micro     55 Standish Crt  Major Office   
4      5        5          Bell Mobility   5099 Creekbank Rd  Major Office   
296  297      373    Royal Bank Building   6880 Financial Dr  Major Office   

     OFF_GFA_m2     OFF_GFAft2  Storeys Year_Built  \
582    30678.35  330219.010927      6.0       2014   
54     32456.00  349353.476267     12.0       1986   
283    43829.00  471771.429360     12.0       1998   
4      46097.00  496183.978170      6.0       1999   
296    64823.10  697750.040898      8.0       1999   

                            geometry  
582  MULTIPOINT (-79.68241 43.62544)  
54   MULTIPOINT (-79.64469 43.58898)  
283  MULTIPOINT (-79.68487 43.62540)  
4    MULTIPOINT (-79.61155 43.64596)  
296  MULTIPOINT (-79.74498 43.60853)

When sorted by `OFF_GFA_m2`, we can see that 33 Queen St South and 6880 Financial Drive have the smallest and largest values respectively. The first looks like a home that was converted into an office, and the second is RBC's Mississauga Office. Looking at the great difference in size, I believe that this field does in fact relate to the size of the office. <br/><br/>
<img src="RefImages/33QueenStSouth.png" width='500' /> <img src="RefImages/6880FinancialDr.png" width="500"/><br><br>
We must now find what the cut off is for high a low density office space. Let's take a look at the distribution of the `OFF_GFA_m2` field.

In [29]:
size_distrib = df_office_bldgs['OFF_GFA_m2'].describe(percentiles=[.25, .38, .4, .42, .5, .75, .9])
size_distrib

count      628.000000
mean      4737.532081
std       6806.174856
min         52.000000
25%        423.000000
38%        897.695800
40%       1069.980000
42%       1221.241200
50%       1792.000000
75%       6531.882500
90%      12825.964000
max      64823.100000
Name: OFF_GFA_m2, dtype: float64

An office of size 1069 m2 looks like a good place to start the search. Looking around the 40th percentile we see the following offices:

In [44]:
tdf = df_office_bldgs[df_office_bldgs.apply(lambda row: row['OFF_GFA_m2'] <= size_distrib['42%'] and row['OFF_GFA_m2'] >= size_distrib['38%'], axis=1)]
tdf.sample(n=10)

FID  BLDG_ID                    BLDG_Name               Address  \
397  398      485      Snap-On Tools of Canada      2325 Skymark Ave   
7      8        8                                    6465 Airport Rd   
444  445      548                                 6800 Campobello Rd   
609  610      735  Port Credit Wellness Centre    333 Lakeshore Rd W   
143  144      185   The Offices at Meadowpines  2420 Meadowpine Blvd   
241  242      301             175 Traders Blvd    175 Traders Blvd E   
443  444      546                                   3465 Semenyk Crt   
442  443      545                                   3464 Semenyk Crt   
330  331      414                                  3095 Glen Erin Dr   
373  374      460                                      29 Tannery St   

           OfficeType  OFF_GFA_m2    OFF_GFAft2  Storeys Year_Built  \
397  Secondary Office     1028.43  11069.928383      1.0       1989   
7    Secondary Office     1161.00  12496.899986      2.0       1973   
444  Secondary Office     1180.00  12701.414284      2.0       1982   
609  Secondary Office     1003.32  10799.646593      2.0       2012   
143  Secondary Office     1070.00  11517.384139      2.0       1990   
241  Secondary Office      929.00   9999.672771      1.0       1986   
443  Secondary Office     1210.80  13032.942724      1.0              
442  Secondary Office     1017.60  10953.355233      2.0              
330  Secondary Office     1009.00  10860.785604      1.0              
373  Secondary Office     1219.82  13130.033196      3.0       2008   

                            geometry  
397  MULTIPOINT (-79.60495 43.65371)  
7    MULTIPOINT (-79.62780 43.69341)  
444  MULTIPOINT (-79.72918 43.60844)  
609  MULTIPOINT (-79.59469 43.54309)  
143  MULTIPOINT (-79.77418 43.60448)  
241  MULTIPOINT (-79.67112 43.62666)  
443  MULTIPOINT (-79.65258 43.56889)  
442  MULTIPOINT (-79.65340 43.56807)  
330  MULTIPOINT (-79.67515 43.53326)  
373  MULTIPOINT (-79.71514 43.58092)

In [51]:
tdf = df_office_bldgs[df_office_bldgs.apply(lambda row: row['OFF_GFA_m2'] >= 1792 and row['OFF_GFA_m2'] <= 6531, axis=1)]
tdf.sample(n=10)

FID  BLDG_ID                           BLDG_Name                Address  \
108  109      143        Mississauga Corporate Centre     3660 Hurontario St   
191  192      241  South Peel Medical Dental Building         77 Queensway W   
284  285      355              Pepsi-Cola Head Office      5205 Satellite Dr   
372  373      459                                        7150 Derrycrest  Dr   
219  220      271     Skymark/Orbitor Place - Phs III       2425 Skymark Ave   
233  234      288                    Queensway Centre      2345 Stanfield Rd   
291  292      364                       Group Telecom         7111 Syntex Dr   
534  535      653                                       2360 Meadowpine Blvd   
97    98      129                                            6870 Goreway Dr   
450  451      555                                      6610 Turner Valley Rd   

           OfficeType  OFF_GFA_m2    OFF_GFAft2  Storeys Year_Built  \
108  Secondary Office     5373.00  57834.490633     10.0       1987   
191  Secondary Office     2033.00  21883.029864      3.0       1975   
284  Secondary Office     5718.73  61555.897379      2.0       1997   
372  Secondary Office     5599.94  60277.252461      3.0  2008/2009   
219  Secondary Office     4213.00  45348.354557      2.0       1989   
233  Secondary Office     2787.00  29999.018313      4.0       1970   
291  Secondary Office     3151.35  33920.849071      3.0       1999   
534  Secondary Office     4324.40  46547.454177      1.0       2010   
97   Secondary Office     2323.00  25004.563882      2.0       1989   
450  Secondary Office     2370.00  25510.467672      2.0       1980   

                            geometry  
108  MULTIPOINT (-79.63389 43.59218)  
191  MULTIPOINT (-79.61001 43.57279)  
284  MULTIPOINT (-79.60871 43.65597)  
372  MULTIPOINT (-79.71215 43.64643)  
219  MULTIPOINT (-79.60248 43.65575)  
233  MULTIPOINT (-79.58315 43.59883)  
291  MULTIPOINT (-79.75635 43.60666)  
534  MULTIPOINT (-79.77210 43.60467)  
97   MULTIPOINT (-79.62790 43.71113)  
450  MULTIPOINT (-79.74138 43.59731)

Anything less than the 25th percentile of 423 will be deemed low density, anything above but under 897 will be medium density, anything higher will be high density. 